File: tree/tree.cpp
#include<iostream>
using namespace std;

class node{
public:
    int data;
    node* left;
    node* right;
}*root;

class link{
public:
    node* data;
    link* next;
}*q;

class queue{
public:
    link* create_link(node* data){
        link *temp=new link;
        temp->data=data;
        temp->next=NULL;
        return temp;
    }
    void enqueue(node* data){
        link *temp,*ptr;
        temp=create_link(data);
        if(q==NULL){
            q=temp;
            q->next=NULL;
            // cout<<"added as root ";
        }
        else{
            ptr=q;
            while(ptr->next!=NULL)
                ptr=ptr->next;
            ptr->next=temp;
            // cout<<"added as link ";
        }
        return;
    }
    node* dequeue(void){
        node* data=q->data;
        q=q->next;
        // cout<<"dequed ";
        return data;
    }
    int len(void){
        link *ptr=q;
        int count=0;
        while(ptr!=NULL){
            ptr=ptr->next;
            count+=1;
        }
        return count;
    }
};

class tree{
public:
    node* create_node(int data){
        node* temp;
        temp=new node;
        temp->data=data;
        temp->left=NULL;
        temp->right=NULL;
        return temp;
    }
    void insert(int data){
        node *temp,*temp1;
        temp1=create_node(data);
        if(root==NULL){
            root=temp1;
            cout<<"added as root node "<<data<<endl;
            return;
        }
        else{
            queue q;
            q.enqueue(root);
            while(q.len()!=0){
                temp=q.dequeue();
                if(temp->left==NULL){
                    temp->left=temp1;
                    cout<<"added as left node "<<data<<endl;
                    break;
                }
                else
                    q.enqueue(temp->left);
                if(temp->right==NULL){
                    temp->right=temp1;
                    cout<<"added as right node "<<data<<endl;
                    break;
                }
                else
                    q.enqueue(temp->right);
            }
        }
        return;
    }
    void inorder(node* ptr){
        if(ptr){
            inorder(ptr->left);
            cout<<ptr->data<<" ";
            inorder(ptr->right);
        }
    }
    void preorder(node* ptr){
        if(ptr!=NULL){
            cout<<ptr->data;
            preorder(ptr->left);
            preorder(ptr->right);
        }
    }

    void postorder(node* ptr){
        if(ptr!=NULL){
            postorder(ptr->left);
            postorder(ptr->right);
            cout<<ptr->data;
        }
    }

    void delet(node* nod){
        node *temp;
        temp=root;
        while(temp->right!=NULL)
            temp=temp->right;
        nod->data=temp->data;
        delete temp;
        cout<<endl<<"delete success"<<endl;
        return;
    }

};
int main(){
    tree tr;
    tr.insert(1);
    tr.insert(2);
    tr.insert(3);
    tr.insert(4);
    tr.insert(5);
    tr.insert(6);
    tr.insert(7);
    cout<<endl<<"inorder :";
    tr.inorder(root);
    cout<<endl<<"preorder :";
    tr.preorder(root);
    tr.delet(root->left);
    cout<<"success";
    cout<<endl<<"preorder :";
    tr.preorder(root);
    tr.insert(8);
    cout<<endl<<"preorder :";
    tr.preorder(root);
    cout<<endl<<"postorder :";
    tr.postorder(root);
    return 0;
}

File: tree/tree.py
class node:
    def __init__(self):
        self.data=None
        self.left=None
        self.right=None


class tree:
    def __init__(self):
        self.root=None

    def add_lnode(self,nod,data):
        temp=node()
        temp.data=data
        nod.left=temp
        print("added left")

    def add_rnode(self,nod,data):
        temp=node()
        temp.data=data
        nod.right=temp
        print("added right")
    
    def insert(self,nod,data):
        q=[]
        q.append(nod)
        while (len(q)):
            temp=q[0]
            q.pop(0)
            if not temp.left:
                self.add_lnode(temp,data)
                break
            else:
                q.append(temp.left)
            if not temp.right:
                self.add_rnode(temp,data)
                break
            else:
                q.append(temp.right)

    def add_node(self,data):
        temp=node()
        temp.data=data
        if self.root == None:
            self.root=temp
            print("added as root")
        else:
            temp1=self.root
            while temp1:
                if temp1.left==None:
                    temp1.left=temp
                    print("added as left")
                    return
                elif temp1.right==None:
                    temp1.right=temp
                    print("added as right")
                    return
                temp1=temp1.left
                print("new node")

    def disp_inorder(self,node):
        if node:
            self.disp_inorder(node.left)
            print(node.data)
            self.disp_inorder(node.right)

    def post_order(self,node):
        if node:
            self.post_order(node.left)
            self.post_order(node.right)
            print(node.data)
            
    def pre_order(self,node):
        if node:
            print(node.data)
            self.pre_order(node.left)
            self.pre_order(node.right)

tr=tree()
tr.add_node(5)
tr.add_node(4)
tr.add_node(3)
tr.add_node(2)
tr.add_lnode(tr.root.right,1)
tr.insert(tr.root,0)
tr.insert(tr.root,6)
print("inorder")
tr.disp_inorder(tr.root)
print("post-order")
tr.post_order(tr.root)
print("pre-order")
tr.pre_order(tr.root)